### PR TITLE
Feature/89 password reset

### LINKED ISSUE #89: Password reset
# User Story

### As a: 
User

### I want:
a way to reset my password if i can't remember

### so that:
i am able to use my account again

## Acceptance criteria:
- [x] Reset link in login page 
- [x] send mail to user with reset link
- [x] reset link referes to a page where the user can set a new password
- [x] invalidate reset link if the user login succesfully / after given time (24h ??)
- [x] allow admins to change passwords of all users

### PR DESCRIPTION
**Pull request by khase**
# Request Title:
Does not implement all improvements mentioned in #89 but adresses the most important ones.

## Description:
 Reset link in login page
 send mail to user with reset link
 reset link referes to a page where the user can set a new password
 invalidate reset link if the user login succesfully / after given time (24h ??)

### CODE REVIEW

File: api/src/controllers/AuthController.ts [hunk 1]
```diff
@@ -36,15 +36,14 @@ export class AuthController {
       })
       .then((savedUser) => {
         emailService.sendActivation(savedUser);
```
Comment (grosspersky) [inline api/src/controllers/AuthController.ts]: The promise doesn't get handled correctly at this position.
So errors in configuration do not get noticed easily.
Some wrong configuration gets me this error:
```
error = Error
 code = "ESOCKET"
 command = "CONN"
 message = "self signed certificate in certificate chain"
 stack = "Error: self signed certificate in certificate chain\n    at TLSSocket.<anonymous> (_tls_wrap.js:1088:38)\n    at emitNone (events.js:86:13)\n    at TLSSocket.emit (events.js:188:7)\n    at TLSSocket._finishInit (_tls_wrap.js:610:8)\n    at TLSWrap.ssl.onhandshakedone (_tls_wrap.js:440:38)"
 __proto__ = Error
```

Added password and user from debugmail.io into main.ts
Provide correct promise handling at this position.

File: api/src/controllers/AuthController.ts [hunk 2]
```diff
@@ -56,10 +55,53 @@ export class AuthController {
...
+        const expires = new Date();
+        expires.setTime((new Date()).getTime()
+          // Add 24h
+          + (24 * 60 * 60 * 1000));
+
+        existingUser.resetPasswordExpires = expires;
+        return existingUser.save();
+      })
+      .then((user) => {
+        emailService.sendPasswordReset(user);
```
Comment (grosspersky) [inline api/src/controllers/AuthController.ts]: Same problem with promises here, as in line 38

**Comment by khase** [comment]
No, you won't have to you any special E-Mail addresses when using debugmail. You actually don#t even have to use valid adresses. Debugmail acts as an dummy smtp server but dont really deliveres your mails. Just look in your debugmail "inbox" and you see all mails that would have beed delivered through a real smtp server.

**Comment by khase** [comment]
I am not able to recrate your exact errors but i got similar ones when providing wrong credentials for the debugmail.io service.

I did just overwrite the following environment variables:
```
      - "MAILPROVIDER=debugmail"
      - "MAILUSER=someUsername"
      - "MAILPASS=aValidPassword"
```

**Comment by grosspersky** [comment]
After update it is at least falling with some better error message:
```
::ffff:127.0.0.1 - - [02/Oct/2017:16:29:16 +0000] "POST /api/auth/register HTTP/1.1" 404 31 "http://localhost:4200/register" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/61.0.3163.100 Safari/537.36"
{ Error: self signed certificate in certificate chain
    at TLSSocket.<anonymous> (_tls_wrap.js:1088:38)
    at emitNone (events.js:86:13)
    at TLSSocket.emit (events.js:188:7)
    at TLSSocket._finishInit (_tls_wrap.js:610:8)
    at TLSWrap.ssl.onhandshakedone (_tls_wrap.js:440:38) code: 'ESOCKET', command: 'CONN' }
(node:1808) UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 1): InternalServerError: Could not send E-Mail
(node:1808) DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
```
Is there any other variable to ignore self signed certificates?

**Pull request merged**